File: Sources/Sentry/SentryMsgPackSerializer.m
#import "SentryMsgPackSerializer.h"
#import "SentryInternalDefines.h"
#import "SentryLogC.h"

@implementation SentryMsgPackSerializer

+ (BOOL)serializeDictionaryToMessagePack:
            (NSDictionary<NSString *, id<SentryStreamable>> *)dictionary
                                intoFile:(NSURL *)path
{
    NSOutputStream *outputStream = [[NSOutputStream alloc] initWithURL:path append:NO];
    [outputStream open];

    uint8_t mapHeader = (uint8_t)(0x80 | dictionary.count); // Map up to 15 elements
    [outputStream write:&mapHeader maxLength:sizeof(uint8_t)];

    for (NSString *key in dictionary) {
        id<SentryStreamable> value = dictionary[key];

        NSData *keyData = [key dataUsingEncoding:NSUTF8StringEncoding];
        uint8_t str8Header = (uint8_t)0xD9; // String up to 255 characters
        uint8_t keyLength = (uint8_t)keyData.length;
        [outputStream write:&str8Header maxLength:sizeof(uint8_t)];
        [outputStream write:&keyLength maxLength:sizeof(uint8_t)];

        [outputStream write:keyData.bytes maxLength:keyData.length];

        NSInteger dataLength = [value streamSize];
        if (dataLength <= 0) {
            // MsgPack is being used strictly for session replay.
            // An item with a length of 0 will not be useful.
            // If we plan to use MsgPack for something else,
            // this needs to be re-evaluated.
            SENTRY_LOG_ERROR(@"Data for MessagePack dictionary has no content - Input: %@", value);
            return NO;
        }

        uint32_t valueLength = (uint32_t)dataLength;
        // We will always use the 4 bytes data length for simplicity.
        // Worst case we're losing 3 bytes.
        uint8_t bin32Header = (uint8_t)0xC6;
        [outputStream write:&bin32Header maxLength:sizeof(uint8_t)];
        valueLength = NSSwapHostIntToBig(valueLength);
        [outputStream write:(uint8_t *)&valueLength maxLength:sizeof(uint32_t)];

        NSInputStream *_Nullable nullableInputStream = [value asInputStream];
        if (nullableInputStream == nil) {
            SENTRY_LOG_ERROR(@"Could not get input stream - Input: %@", value);
            [outputStream close];
            return NO;
        }
        NSInputStream *_Nonnull inputStream
            = SENTRY_UNWRAP_NULLABLE(NSInputStream, nullableInputStream);
        [inputStream open];

        uint8_t buffer[1024];
        NSInteger bytesRead;

        while ([inputStream hasBytesAvailable]) {
            bytesRead = [inputStream read:buffer maxLength:sizeof(buffer)];
            if (bytesRead > 0) {
                [outputStream write:buffer maxLength:bytesRead];
            } else if (bytesRead < 0) {
                SENTRY_LOG_ERROR(@"Error reading bytes from input stream - Input: %@ - %li", value,
                    (long)bytesRead);

                [inputStream close];
                [outputStream close];
                return NO;
            }
        }

        [inputStream close];
    }
    [outputStream close];

    return YES;
}

@end

@implementation NSURL (SentryStreameble)

- (nullable NSInputStream *)asInputStream
{
    return [[NSInputStream alloc] initWithURL:self];
}

- (NSInteger)streamSize
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSError *error;
    NSString *_Nullable nullablePath = self.path;
    if (nullablePath == nil) {
        SENTRY_LOG_DEBUG(@"File URL has no path - File: %@", self);
        return -1;
    }
    NSDictionary *attributes =
        [fileManager attributesOfItemAtPath:SENTRY_UNWRAP_NULLABLE(NSString, nullablePath)
                                      error:&error];
    if (attributes == nil) {
        SENTRY_LOG_DEBUG(@"Could not read file attributes - File: %@ - %@", self, error);
        return -1;
    }
    NSNumber *fileSize = attributes[NSFileSize];
    return [fileSize unsignedIntegerValue];
}

@end

@implementation NSData (SentryStreameble)

- (nullable NSInputStream *)asInputStream
{
    return [[NSInputStream alloc] initWithData:self];
}

- (NSInteger)streamSize
{
    return self.length;
}

@end

File: Sources/Swift/Tools/MsgPack/SentryMsgPackSerializer.swift
/**
 * This is a partial implementation of the MessagePack format.
 * We only need to concatenate a list of NSData into an envelope item.
 */
final class SentryMsgPackSerializer {
    static func serializeDictionary(toMessagePack dictionary: [String: SentryStreamable], intoFile fileURL: URL) -> Bool {
        do {
            try serializeToFile(dictionary: dictionary, fileURL: fileURL)
            return true
        } catch {
            SentrySDKLog.error("Failed to serialize dictionary to MessagePack - Error: \(error)")
            // Clean up partial file on error
            do {
                try FileManager.default.removeItem(at: fileURL)
            } catch {
                // Ignore cleanup errors - file might not exist
            }
            return false
        }
    }
    
    // swiftlint:disable:next function_body_length cyclomatic_complexity
    private static func serializeToFile(dictionary: [String: SentryStreamable], fileURL: URL) throws {
        guard let outputStream = OutputStream(url: fileURL, append: false) else {
            throw SentryMsgPackSerializerError.outputError("Failed to create output stream for file: \(fileURL)")
        }
        outputStream.open()
        defer { 
            outputStream.close()
        }
        
        // Check if stream opened successfully
        if outputStream.streamError != nil {
            throw SentryMsgPackSerializerError.outputError("Failed to open output stream for file: \(fileURL)")
        }
        
        let mapHeader = UInt8(truncatingIfNeeded: 0x80 | dictionary.count) // Map up to 15 elements
        _ = outputStream.write([mapHeader], maxLength: 1)

        for (key, value) in dictionary {
            let keyData = Data(key.utf8)
            
            let str8Header: UInt8 = 0xD9 // String up to 255 characters
            let keyLength = UInt8(truncatingIfNeeded: keyData.count) // Truncates if > 255, matching Objective-C behavior
            _ = outputStream.write([str8Header], maxLength: 1)
            _ = outputStream.write([keyLength], maxLength: 1)
            
            keyData.withUnsafeBytes { bytes in
                guard let bufferAddress = bytes.bindMemory(to: UInt8.self).baseAddress else {
                    return
                }
                _ = outputStream.write(bufferAddress, maxLength: keyData.count)
            }

            guard let dataLength = value.streamSize(), dataLength > 0 else {
                // MsgPack is being used strictly for session replay.
                // An item with a length of 0 will not be useful.
                // If we plan to use MsgPack for something else,
                // this needs to be re-evaluated.
                throw SentryMsgPackSerializerError.emptyData("Data for MessagePack dictionary has no content - Input: \(value)")
            }

            var valueLength = UInt32(truncatingIfNeeded: dataLength)
            // We will always use the 4 bytes data length for simplicity.
            // Worst case we're losing 3 bytes.
            let bin32Header: UInt8 = 0xC6
            _ = outputStream.write([bin32Header], maxLength: 1)
            
            // Write UInt32 as big endian bytes
            valueLength = NSSwapHostIntToBig(valueLength)
            withUnsafeBytes(of: valueLength) { bytes in
                guard let baseAddress = bytes.bindMemory(to: UInt8.self).baseAddress else {
                    return
                }
                _ = outputStream.write(baseAddress, maxLength: 4)
            }

            guard let inputStream = value.asInputStream() else {
                throw SentryMsgPackSerializerError.streamError("Could not get input stream - Input: \(value)")
            }
            
            inputStream.open()
            defer { inputStream.close() }

            var buffer = [UInt8](repeating: 0, count: 1_024)
            var bytesRead: Int

            while inputStream.hasBytesAvailable {
                bytesRead = inputStream.read(&buffer, maxLength: buffer.count)
                if bytesRead > 0 {
                    _ = outputStream.write(buffer, maxLength: bytesRead)
                } else if bytesRead < 0 {
                    throw SentryMsgPackSerializerError.streamError("Error reading bytes from input stream - Input: \(value) - Bytes read: \(bytesRead)")
                }
            }
        }
    }
}
